### DISCUSSION TITLE
More usefull checks?

### DISCUSSION THREAD

**Discussion by jjnxpct** [akeeba/panopticon#678] [Ideas]
Panopticon is capable of managing doing backups, fielscans and extension management. All are great tools!

The external service we use at this moment also does some other useful checks on our sites. Most of these checks probably only need to be check once, to see if a site is using 'sane' security settings. Some other checks might indicate 'bad' user behaviour like uploading huge images of documents.

Some example:

- The Debug Mode Should Be Disabled
- Do Not Use $live_site Configuration
- Error Reporting Should Be Set To None On A Live Site
- Installation Folders Should Be Deleted
- tmp/logs Folders Should Be Writable
- Remove Unneeded Joomla Core "fluff"
- Akeeba Kickstart Should Not Be Left In Webspace
- Super Admin Should Not Have Username "admin"
- Only Enable User Registration If Needed
- Use Multi-Factor Authentication On All Super User Accounts
- Remove "Never Logged In" Accounts
- Check for hidden files
- File permissions
- Locate And Review Files Over 2Mb Size
- Review "Renamed To Hide" Files (like File.old, File.bak)
- PHP Files Should Not Be In These Certain Folders
- Locate And Review Admintool_breaches.log Files
- Identify Files With No Content (Zero Bytes In Size)
- Identify Core Joomla Files That Are Missing From Your Webspace

Maybe it would be a nice feature to have some kind of _library_ of checks rules. And those rules can be applied to the sites so we can audit the sites on those rules. 

One a set of rules could be set to check manually (for example before publishing new website) Some other set of rules might be set to run periodically to keep an eye on some specific things, like big files sizes.

I would be up to you what 'rules' (checks) would help users to harden the site security. And users of Panopticon would have the flexibility to add check rules to their sites. Or not.

What do you think?

**Comment by brianteeman** [comment]
panopticon is not a clone of myjoomla

**Reply by jjnxpct** [reply]
I know. It's an alternative. Why not compare / suggest features?

**Comment by nikosdion** [comment]
Panopticon is site monitoring. It does not do security. That would be Admin Tools Pro, a different product. So, I’ll evaluate your suggestions as Admin Tools features. 

Do note that a lot of the features you mention have long been obsolete, or only ever made sense when performing diagnostics. So, I’ll put them in categories.

Needs external context to make sense:
The Debug Mode Should Be Disabled. Unless you’re troubleshooting. Make it part of your troubleshooting checklist. The presence of the debug bar in any version of Joomla released the last decade is a dead giveaway.
Error Reporting Should Be Set To None On A Live Site. See above. 
Review "Renamed To Hide" Files (like File.old, File.bak). There is no universal naming pattern. In many cases, you NEED to keep these files, e.g. when applying settings you don’t know if they will work. So, without context of why these files are there, this check is more likely to do harm than good. 
Meaningless, or obsolete:
Identify Files With No Content (Zero Bytes In Size). Hard disagree. Several flag files have a legitimate purpose and no content. Remove them and shit will break.
Do Not Use $live_site Configuration. It doesn’t do much harm unless you are really daft about it, e.g. set it to plain HTTP on a site with HSTS enabled. Admin Tools warns you in all common use cases which involve operator daftness. We are keenly ware of the exponential increase in daftness relative to undercaffeination. 
Remove Unneeded Joomla Core "fluff". Has not been present in the core distribution for the most part for the better part of a decade. Inconsequential when using Admin Tools Pro and the .htaccess Maker.
Super Admin Should Not Have Username "admin".  No longer relevant. This was advice from back when 2FA / MFA wasn’t a core feature, i.e. before J3.2.
Locate And Review Files Over 2Mb Size. Pretty pointless in 2024 when people upload media dozens of times larger than this. It would only make sense if you only checked for certain non-media files. It would still be finding a needle in a haystack on sites like ours which delivers software as ZIP files.
Locate And Review Admintool_breaches.log Files. Only generated when explicitly enabled for troubleshooting purposes, then removed. This check is obsolete. It was only relevant up until around seven or so years ago. 
Use Multi-Factor Authentication On All Super User Accounts. I wrote the code in the core which allows you to ENFORCE that. We also do far more serious checks so you don’t get unexpected Super Admin users which is what matters even more. 
Check for hidden files. Hidden files are inconsequential to and often needed, e.g. the entire .well-known folder, various .gitkeep types of files, .user.ini for PHP settings, etc. How many sites broke because people blindly deleted hidden files marked by this feature? Sigh.
Needs to be part of your checklists:
Only Enable User Registration If Needed. Part of your site building checklist.
Installation Folders Should Be Deleted. Part of your site building and restoration checklist. If they’re not, Joomla redirects to them. If you are stupid enough to rename them to something like installation.bak, well, stand 30 yards away from a wall, sprint towards it, and hit it with your head at full speed. Repeat until stupidity is fixed one way or another. You know who you are! I’ve seen your sites!
Akeeba Kickstart Should Not Be Left In Webspace. Part of your site restoration checklist. See notes on installation folder. 
Already implemented:
Remove "Never Logged In" Accounts. Admin Tools can do that. Very bad idea when you apply it as a blanket policy, especially if you’re selling things tied to a user account! You need to carefully evaluate your use case before doing this. 
File permissions. Checking file permissions is dumb because the tool doesn’t know the ownership of the files and the web server. Even if it does, it won’t know any file attributes / Linux ACLs in place. This is why Admin Tools lets you change permissions, not “check” them.
tmp/logs Folders Should Be Writable. Checked by Admin Tools. 
PHP Files Should Not Be In These Certain Folders. Irrelevant when using Admin Tools' .htaccess Maker. 
Already scheduled:
Identify Core Joomla Files That Are Missing From Your Webspace. I’d say 
also tell me which are not pristine. This is already in the Issues tracker of Panopticon for future development.

**Comment by jjnxpct** [comment]
Well, that is actually pretty helpful. Thanks for sharing this Nicholas.  To summarize: Most 'checks' provided by some other services are not relevant (anymore) or should be part of a **site checklist** when creating sites.  Also Admin Tools (also) provides useful checks that do not need to done by external services.

The tricky thing about all these checks is: It's not always that easy for me to evaluate all those checks like you are able to do.  I just want to make sure I do the best I can to keep the sites of our clients as safe as possible, using the right tools and procedures.

**Reply by nikosdion** [reply]
This kind of features is always really hard to parse if you don't have the context of their implementation.

For example, scanning permissions _used_ to make sense. Back then, we'd have shared hosts which were running Apache as `nobody:nogroup`, with PHP as an Apache module, service hundreds of sites. It was... challenging. Then we had suPHP and things got a bit better. Then we had PHP-FPM with individual pools per site, making things a lot better. Now we have VMs and containerization which essentially means each site runs as its own (virtual) computer, relegating ownership and permissions issues to the historical footnote category. Still, you _may_ find a badly configured server, usually a self-hosted affair used by someone who followed a generic guide written by "AI" or a (usually East Asian) noob, replicating the issues of the past. Would I want to waste the CPU power and risk confusion for 99.9% of users who don't have a broken environment? I would say no. Hosted services are usually purchased based on feature count, rather than fit for purpose, so they're likely to say yes.

In that last bit lies a lot of the unspoken truth about why features which no longer make sense are still around in site monitoring services. Even if their owners want to get rid of them (and I assume they do want to get rid of a lot of them) doing so would be business suicide given the utterly wrong way people shop for these services. Panopticon being a free of charge, Free and Open Source, self-hosted _application_ doesn't have to be shackled to these conventions. We can afford to be opinionated.

Also, do remember that about a third of the features are only obsolete because I am evaluating them in the context of you having Admin Tools Pro on your site (since Panopticon isn't a security solution, Admin Tools Pro is). For sites which don't have Admin Tools Pro these features still do make sense.

Finally, a little note about these features and the core file scanner feature. These all work by scanning each and every file of your site. When we implement the core files check, we will most likely have to roll it into Admin Tools Pro's PHP File Change Scanner. Otherwise, we need to copy most of that feature into the Panopticon connector which is possible for Joomla! 4+ but not for Joomla! 3 or WordPress (yes, WP support _is_ coming; I am currently working on it). Since you had a ticket on our business site asking why you should use the PHP File Change Scanner: that would be a very valid reason in your use case.
